### PR TITLE
Implement FirstViewController programmatically

### PR DESCRIPTION
**Pull request by kstoletniy**
#### Description:

* Delete FirstViewController from storyboard
* Implement FirstViewController programmatically
* Define useful resources like colors, localizable strings and font
* Define useful helpers for constraints and constant

---

#### Notes:

* The rest of the viewControllers will be implemented in the next PR

---

#### Tasks:

[ISSUE-177](https://github.com/rootstrap/ios-base/issues/177)

---

#### Preview:

Simulator Screen Shot - iPhone Xs Max - 2021-07-13 at 13 41 17

### CODE REVIEW

File: ios-base/Extensions/ColorExtension.swift [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+//  ColorExtension.swift
+//  ios-base
+//
+//  Created by Karen Stoletniy on 12/7/21.
+//  Copyright © 2021 Rootstrap Inc. All rights reserved.
+//
+
+import UIKit
+
+// swiftlint:disable force_unwrapping
```
Comment (glm4) [inline ios-base/Extensions/ColorExtension.swift:11]: Please re enable the rule after the extension declaration or use `disable:next`

File: ios-base/Extensions/ButtonExtension.swift [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+extension UIButton {
+  
+  static func primaryButton(
+    color: UIColor = .buttonBackground,
+    title: String = "",
+    titleColor: UIColor = .white,
+    cornerRadius: CGFloat = UI.Button.cornerRadious,
+    height: CGFloat = UI.Button.height,
+    width: CGFloat = UI.Button.width,
```
Comment (germanStabile) [inline ios-base/Extensions/ButtonExtension.swift:20]: you can remove width as its not being used

File: ios-base/Extensions/FontExtension.swift [hunk 3]
```diff
@@ -10,9 +10,12 @@ import UIKit
...
 extension UIFont {
-  var bold: UIFont { return withWeight(.bold) }
-  var semibold: UIFont { return withWeight(.semibold) }
-  var light: UIFont { return withWeight(.ultraLight) }
+  static let h1Regular: UIFont = .systemFont(ofSize: 32, weight: .regular)
+  static let h2Regular: UIFont = .systemFont(ofSize: 16, weight: .regular)
+  static let h3Regular: UIFont = .systemFont(ofSize: 15, weight: .regular)
+  static let h1Medium: UIFont = .systemFont(ofSize: 32, weight: .medium)
+  static let h2Medium: UIFont = .systemFont(ofSize: 16, weight: .medium)
+  static let h3Medium: UIFont = .systemFont(ofSize: 15, weight: .medium)
```
Comment (germanStabile) [inline ios-base/Extensions/FontExtension.swift:18]: shall we use the `font(withName:)` helper here to use proportional sizes?

File: ios-base/Extensions/ViewExtension.swift [hunk 4]
```diff
@@ -53,6 +53,61 @@ extension UIView {
...
+        constant: verticalOffset
+      ).isActive = true
+    }
+    
+    if horizontally {
+      subview.centerXAnchor.constraint(
+        equalTo: centerXAnchor,
+        constant: horizontalOffset
+      ).isActive = true
+    }
```
Comment (glm4) [inline ios-base/Extensions/ViewExtension.swift:109]: this could crash if the view is not added to the hierarchy, which is an assumption this method is doing. Please add a check for that case or add the proper docstrings for the function.

File: ios-base/Extensions/ColorExtension.swift [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+extension UIColor {
```
Comment (germanStabile) [inline ios-base/Extensions/ColorExtension.swift:12]: didn't we have a UIColor extension already?

File: ios-base/Helpers/Constants.swift [hunk 6]
```diff
@@ -27,3 +28,26 @@ enum ErrorDomain: String {
...
+  enum Defaults {
+    static let margin: CGFloat = 32
+  }
+  
+  enum ViewController {
+    static let topMargin: CGFloat = 72
+  }
+  
+  enum Button {
+    static let cornerRadious: CGFloat = 22.0
```
Comment (glm4) [inline ios-base/Helpers/Constants.swift:42]: Please use US English, `cornerRadius`

**Review by germanStabile — APPROVED** [review]
Looks good! Thanks for tackling this 😄

File: ios-base/Onboarding/Routes/OnboardingRoutes.swift [hunk 7]
```diff
@@ -42,9 +42,7 @@ enum OnboardingRoutes: Route {
   }
 
   private func buildFirstViewController() -> UIViewController {
-    guard let first = R.storyboard.main.firstViewController() else {
-      return UIViewController()
-    }
+    let first = FirstViewController()
```
Comment (glm4) [inline ios-base/Onboarding/Routes/OnboardingRoutes.swift:45]: the variable name is not very intuitive nor descriptive.

File: ios-base/Extensions/FontExtension.swift (hunk 3 above)
Comment (kstoletniy) [reply ios-base/Extensions/FontExtension.swift:18]: do we know the font name for .systemFont?

File: ios-base/Extensions/ColorExtension.swift (hunk 5 above)
Comment (kstoletniy) [reply ios-base/Extensions/ColorExtension.swift:12]: no, we didn't

File: ios-base/Extensions/FontExtension.swift (hunk 3 above)
Comment (germanStabile) [reply ios-base/Extensions/FontExtension.swift:18]: you can default the name param to an empty string and it will default to system if the font doesn't exist. Also it would be great to update that func to the EInsights version where we limit the proportional size

Comment (kstoletniy) [reply ios-base/Extensions/FontExtension.swift]: thanks!

File: ios-base/Extensions/ViewExtension.swift [hunk 8]
```diff
@@ -53,6 +53,70 @@ extension UIView {
...
+  
+  /// Center a subview already added to its parent view
+  ///
+  /// - Parameters:
+  ///   - subview: View to be centered on its parent
+  ///   - verticalOffset: Float indicating the vertical displacement of the view
+  ///   - horizontalOffset: Float indicating the horizontal displacement of the view
+  ///   - vertically: Boolean that indicates if the view will be center vertically
+  ///   - horizontally: Boolean that indicates if the view will be center horizontally
+
```
Comment (glm4) [inline ios-base/Extensions/ViewExtension.swift:98]: nitpick: please remove extra line.

File: ios-base/Helpers/Constants.swift [hunk 9]
```diff
@@ -27,3 +28,26 @@ enum ErrorDomain: String {
   case generic = "GenericError"
   case parsing = "ParsingError"
 }
+
+struct UI {
```
Comment (glm4) [inline ios-base/Helpers/Constants.swift:32]: nitpick: Please move this to its own file. This file `Constants` seems to be growing indefinitely with all sort of things.

File: ios-base/Helpers/Constants.swift [hunk 10]
```diff
@@ -27,3 +28,26 @@ enum ErrorDomain: String {
...
+    static let height: CGFloat = 45.0
+    static let width: CGFloat = 200.0
+    static let spacing: CGFloat = 20.0
+  }
+  
+  enum TextField {
+  }
+  
+  enum Label {
+  }
```
Comment (glm4) [inline ios-base/Helpers/Constants.swift:52]: nitpick: Pls remove this if not used

File: ios-base/Extensions/ViewExtension.swift [hunk 11]
```diff
@@ -53,6 +53,70 @@ extension UIView {
...
+      subview.bottomAnchor.constraint(equalTo: bottomAnchor, constant: -bottomMargin)
+    ])
+  }
+  
+  /// Center a subview already added to its parent view
+  ///
+  /// - Parameters:
+  ///   - subview: View to be centered on its parent
+  ///   - verticalOffset: Float indicating the vertical displacement of the view
+  ///   - horizontalOffset: Float indicating the horizontal displacement of the view
```
Comment (glm4) [inline ios-base/Extensions/ViewExtension.swift:95]: Please use the correct type, for UIKit the values are usually CGFloat, and it is also what the function is expecting.

File: ios-base/Extensions/ViewExtension.swift (hunk 4 above)
Comment (glm4) [reply ios-base/Extensions/ViewExtension.swift:118]: I think a better approach would be making it the other way around. Like:
```
func centerInView(parent: UIView, ...) {
  if parent != superview {
    parent.addSubview(self)
  } 
  ...
}
```

This way you also avoid the issues with hierarchy.

File: ios-base/Extensions/FontExtension.swift [hunk 12]
```diff
@@ -29,11 +32,17 @@ extension UIFont {
...
-    let size = Font.PointSize.proportional(to: (.screen6_5Inch,
-                                                size)).value()
+  static func font(withName name: String = "", size: CGFloat) -> UIFont {
+    let minSize = size - 1
+    let maxSize = size + 2
+
+    let proportionalSize = Font.PointSize
+      .proportional(to: (.screen4_7Inch, size)).value()
+
+    let finalSize = max(min(proportionalSize, maxSize ), minSize)
```
Comment (glm4) [inline ios-base/Extensions/FontExtension.swift:42]: please revert these changes. The issue with bigger fonts is not related to the ios-base, it is a known issue with RSFontSizes and the Device dependency being updated. We need to update the dependency and then update the pod version of RSFontSizes.

**Comment by glm4** [comment]
@kstoletniy Please add the issue that is being fixed to the description of the PR.

**Comment by kstoletniy** [comment]
> @kstoletniy Please add the issue that is being fixed to the description of the PR.

@glm4 it's already added under Task section

**Comment by glm4** [comment]
> > @kstoletniy Please add the issue that is being fixed to the description of the PR.
> 
> @glm4 it's already added under Task section

I missed it, my bad. I guess that we are not using the right PR template 😬

File: ios-base/Helpers/Constants.swift (hunk 10 above)
Comment (kstoletniy) [reply ios-base/Helpers/Constants.swift:52]: will be used in the followings PR when the rest of the views are to be implemented

**Comment by kstoletniy** [comment]
@glm4 sorry, i don't get the comment https://github.com/rootstrap/ios-base/pull/180#discussion_r669065596
what is the point to check the parent view and add it to the current view?

What i think that we need to check if the subview to be centred is added to the view, like this:
```
    if !subviews.contains(subview) {
      addSubview(subview)
    }
```

**Comment by kstoletniy** [comment]
@glm4 sorry, i don't get the comment #180 (comment)
what is the point to check the view has a parent view and add it if is missing?

What i think that we need to check if the subview to be centred is added to the view, like this:

```
    if !subviews.contains(subview) {
      addSubview(subview)
    }
```

File: ios-base/Helpers/Constants.swift (hunk 10 above)
Comment (glm4) [reply ios-base/Helpers/Constants.swift:52]: The overall recommendation is to keep the codebase as clean as possible and only add code that is used in the current version branch. This can be easily be added when it is needed since now it serves no purpose and have no members.

**Comment by glm4** [comment]
> @glm4 sorry, i don't get the comment #180 (comment)
> what is the point to check the view has a parent view and add it if is missing?
> 
> What i think that we need to check if the subview to be centred is added to the view, like this:
> 
> ```
>     if !subviews.contains(subview) {
>       addSubview(subview)
>     }
> ```

For better tracking I will reply this in the comment thread. Please do the same we can continue the conversation there.

**Comment by kstoletniy** [comment]
> > @glm4 sorry, i don't get the comment #180 (comment)
> > what is the point to check the view has a parent view and add it if is missing?
> > What i think that we need to check if the subview to be centred is added to the view, like this:
> > ```
> >     if !subviews.contains(subview) {
> >       addSubview(subview)
> >     }
> > ```
> 
> For better tracking I will reply this in the comment thread. Please do the same we can continue the conversation there.

git bug 😂 it didn't allow me to follow in the thread

File: ios-base/Extensions/ViewExtension.swift (hunk 4 above)
Comment (glm4) [reply ios-base/Extensions/ViewExtension.swift:118]: @kstoletniy please see my method's signature. To this function you pass the parent view where you want to center the current instance. It makes more sense since you are calling this function on that instance, not the parent.
i.e `subview.centerInView(parent, ...)`

The point of checking for the parent and adding self as a subview is that you don't trust the caller to read the docs and expect them to crash, also you don't assume they will have the correct view hierarchy.

On a side note, having control paramenters (`vertically`) is a code smell. This can be split in two functions like:
`centerVertically(inParent parent: UIView, ...)` and `centerHorizontally(inParent parent: UIView, ...)`.

File: ios-base/Extensions/FontExtension.swift (hunk 12 above)
Comment (germanStabile) [reply ios-base/Extensions/FontExtension.swift:42]: shall we leave a TODO until the RSFintSizes issue is fixed?

File: ios-base/Extensions/ViewExtension.swift (hunk 4 above)
Comment (germanStabile) [reply ios-base/Extensions/ViewExtension.swift:118]: I think it goes better as is with the rest of the auto layout helpers:

```
  superView.attachHorizontally(subview)
  superView.activateMarginConstraints(n: subview)
```

also I don't see how having control parameters is a code smell? 
but anyways we can update so we have
```
centerVertically()
centerHorizontally()

func center(subview: UIView) {
  centerVertically()
  centerHorizontally()
}

Comment (glm4) [reply ios-base/Extensions/ViewExtension.swift:121]: IMO the latter reads much better, and the call it is directly related to the current view instance which makes clear the result.
You can read more about control parameters here: https://martinfowler.com/bliki/FlagArgument.html

File: ios-base/Extensions/ButtonExtension.swift [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+      cornerRadius: cornerRadius,
+      height: height,
+      font: font
+    )
+    if let action = action {
+      button.addTarget(
+        target,
+        action: action,
+        for: .touchUpInside
+      )
```
Comment (glm4) [inline ios-base/Extensions/ButtonExtension.swift:38]: please don't split lines if not needed

File: ios-base/Extensions/ColorExtension.swift [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+//  ColorExtension.swift
+//  ios-base
+//
+//  Created by Karen Stoletniy on 12/7/21.
+//  Copyright © 2021 Rootstrap Inc. All rights reserved.
+//
+
+import UIKit
+
+extension UIColor {
```
Comment (glm4) [inline ios-base/Extensions/ColorExtension.swift:11]: @kstoletniy  did you disable the `force_unwrapping` rule all together?

File: ios-base/Extensions/FontExtension.swift [hunk 15]
```diff
@@ -10,9 +10,12 @@ import UIKit
...
 extension UIFont {
-  var bold: UIFont { return withWeight(.bold) }
-  var semibold: UIFont { return withWeight(.semibold) }
-  var light: UIFont { return withWeight(.ultraLight) }
+  static let h1Regular: UIFont = .font(size: 32).withWeight(.regular)
+  static let h2Regular: UIFont = .font(size: 16).withWeight(.regular)
+  static let h3Regular: UIFont = .font(size: 15).withWeight(.regular)
+  static let h1Medium: UIFont = .font(size: 32).withWeight(.medium)
+  static let h2Medium: UIFont = .font(size: 16).withWeight(.regular)
+  static let h3Medium: UIFont = .font(size: 15).withWeight(.regular)
```
Comment (glm4) [inline ios-base/Extensions/FontExtension.swift:18]: Please move the font sizes to meaningful constants as well.

File: ios-base/Helpers/Constants.swift [hunk 16]
```diff
@@ -7,6 +7,7 @@
 //
 
 import Foundation
+import UIKit
```
Comment (glm4) [inline ios-base/Helpers/Constants.swift:10]: Pls remove if not needed anymore

File: ios-base/Extensions/ViewExtension.swift (hunk 4 above)
Comment (FLTravieso) [reply ios-base/Extensions/ViewExtension.swift]: I'm agree with @glm4. It's more readable in that way so I've done a refactor for this methods and attach methods as well. However I didn't add the `addSubview` statement. If the method is called with a wrong view then we add it and center it would end on a misplaced behavior for the caller. Please check the new implementation and let me know your thoughts.

File: ios-base/Extensions/ViewExtension.swift [hunk 17]
```diff
@@ -55,70 +55,76 @@ extension UIView {
...
-    if !subviews.contains(subview) {
-      addSubview(subview)
-    }
-    
-    if vertically {
-      subview.centerYAnchor.constraint(
-        equalTo: centerYAnchor,
-        constant: verticalOffset
-      ).isActive = true
+    guard superview == view else {
```
Comment (germanStabile) [inline ios-base/Extensions/ViewExtension.swift:99]: this isn't needed IMO. you may want to center sibling views

File: ios-base/Extensions/ViewExtension.swift [hunk 18]
```diff
@@ -55,70 +55,76 @@ extension UIView {
...
+  /// Centers the view vertically  in a specific view
+  ///
+  /// - Parameters:
+  ///   - view: UIView on which the view will be centered vertically
+  ///   - withOffset: CGFloat indicating the vertical displacement of the view
+  func centerVerticallyIn(
+    _ view: UIView,
+    withOffset offset: CGFloat = 0
+  ) {
+    guard superview == view else {
```
Comment (germanStabile) [inline ios-base/Extensions/ViewExtension.swift:119]: same

File: ios-base/Extensions/FontExtension.swift (hunk 15 above)
Comment (FLTravieso) [reply ios-base/Extensions/FontExtension.swift:18]: I don't like to have another enum with more constants. We got almost all of them added in `Font.PointSize`. I'm wondering if wouldn't be better to modify the `RSFontSizes` library to use `proportional` as a function instead a case so we can use `Font.PointSize` constants (and the case `fixed` for the not added ones)  and ask then for the proportion as we seem to need in the line 36

File: ios-base/Extensions/ViewExtension.swift (hunk 17 above)
Comment (germanStabile) [reply ios-base/Extensions/ViewExtension.swift:99]: please add

```
func center(_ view: UIView, withOffset offset: CGPoint = .zero) {
   centerHorizontallyIn(view, withOffset: offset.x)
   centerVerticallyIn(view, withOffset: offset.y)
}

File: ios-base/Extensions/ViewExtension.swift (hunk 18 above)
Comment (germanStabile) [reply ios-base/Extensions/ViewExtension.swift:119]: name should be:

```
func centerVertically(
     with view: UIView,
     withOffset offset: CGFloat = 0
```
same for the horizontal counterpart

File: ios-base/Extensions/ViewExtension.swift (hunk 17 above)
Comment (FLTravieso) [reply ios-base/Extensions/ViewExtension.swift:99]: > this isn't needed IMO. you may want to center sibling views

We need this to be sure that the view is already added to that view. We could change this to check if the view is a subview of the view that's coming through param otherwise this will crash.

Another option is to chang the method to be `centerHorizontallyInSuperview` then no have `view` parameter and just check inside if `superview` is not nil

Comment (germanStabile) [reply ios-base/Extensions/ViewExtension.swift:99]: you may have View A and B, both subviews of View C. And you may want to center view A with View B.
This would be possible without the current check. I think its responsibility of the client to see that both View A and B are in the same view hierarchy.

File: ios-base/Extensions/ViewExtension.swift (hunk 18 above)
Comment (FLTravieso) [reply ios-base/Extensions/ViewExtension.swift:119]: I think you meant:

```suggestion
    func centerVertically(
     in view: UIView,
     withOffset offset: CGFloat = 0
```

If we decide to take the last approach I metioned [here](https://github.com/rootstrap/ios-base/pull/180#discussion_r681196414) this is not needed

Comment (germanStabile) [reply ios-base/Extensions/ViewExtension.swift:119]: Please see my last reply. by doing so you don't necessarily have one view being superview of the other as long as they are in the same view hierarchy, so the word `with` is more appropriate

File: ios-base/Extensions/FontExtension.swift (hunk 15 above)
Comment (glm4) [reply ios-base/Extensions/FontExtension.swift:18]: We shouldn't modify a dependency for something that is convenient to us. The purpose of proportional is to provide a flexible way of scaling the base point size passed by parameter. `Font.PointSize`  are simply to provide meaningful predefined sizes but doesn't mean that we can have our own PointSize system, either an enum, constants or any other.

File: ios-base/Extensions/ViewExtension.swift [hunk 19]
```diff
@@ -53,6 +53,79 @@ extension UIView {
...
+    centerVertically(with: view, withOffset: offset.y)
+  }
+
+  /// Centers the view horizontally with a specific view
+  ///
+  /// - Parameters:
+  ///   - view: UIView on which the view will be centered horizontally
+  ///   - withOffset: CGFloat indicating the horizontal displacement of the view
+  func centerHorizontally(
+    with view: UIView,
```
Comment (glm4) [inline ios-base/Extensions/ViewExtension.swift:106]: maybe `on`  is a better preposition

File: ios-base/Extensions/ViewExtension.swift [hunk 20]
```diff
@@ -53,6 +53,79 @@ extension UIView {
...
+    ).isActive = true
+  }
+
+  /// Centers the view vertically with a specific view
+  ///
+  /// - Parameters:
+  ///   - view: UIView on which the view will be centered vertically
+  ///   - withOffset: CGFloat indicating the vertical displacement of the view
+  func centerVertically(
+    with view: UIView,
```
Comment (glm4) [inline ios-base/Extensions/ViewExtension.swift:121]: same here, `on view:` fits better IMO.

**Review by glm4 — APPROVED** [review]

File: ios-base/Extensions/ViewExtension.swift (hunk 19 above)
Comment (germanStabile) [reply ios-base/Extensions/ViewExtension.swift:106]: I don't mind either way but I recommended `with` since its not necessary that one view contains the other.

**Review by pMalvasio — APPROVED** [review]
